pkg/aws/ec2/ec2_test.go

```diff
@@ -379,3 +300,208 @@ func TestCollector_Collect(t *testing.T) {
+
+			ps.EXPECT().
+				GetProducts(mock.Anything, mock.Anything, mock.Anything).
+				RunAndReturn(tt.GetProducts).
+				Times(tt.expectedCalls)
+
+			err := collector.populateStoragePricingMap()
+			if tt.err != nil {
+				assert.ErrorIs(t, err, tt.err)
+			}
+			assert.Equal(t, tt.expected, collector.storagePricingMap)
+			assert.Equal(t, tt.expectedScrapeTimeCompare, collector.StorageScrapingInterval.Compare(scrapingTimeAtCreation))
+		})
+	}
+}
+
+func Test_FetchVolumesData(t *testing.T) {
+	t.Run("sends EBS volumes data to channel", func(t *testing.T) {
+		regionName := "af-south-1"
+		region := ec2Types.Region{
+			RegionName: aws.String(regionName),
+		}
+
+		ps := mockpricing.NewPricing(t)
+		collector := New(&Config{
+			Regions: []ec2Types.Region{region},
+			Logger:  logger,
+		}, ps)
+
+		client := mockec2.NewEC2(t)
+		client.EXPECT().
+			DescribeVolumes(mock.Anything, mock.Anything, mock.Anything).
+			RunAndReturn(
+				func(ctx context.Context, e *ec2.DescribeVolumesInput, optFns ...func(*ec2.Options)) (*ec2.DescribeVolumesOutput, error) {
+					return &ec2.DescribeVolumesOutput{
+						Volumes: []ec2Types.Volume{
+							{
+								VolumeId: aws.String("vol-111111111"),
+							},
+						},
+					}, nil
+				},
+			).
+			Times(1)
+
+		wg := sync.WaitGroup{}
+		wg.Add(len(collector.Regions))
+		ch := make(chan []ec2Types.Volume)
+		go collector.fetchVolumesData(client, regionName, &wg, ch)
+		go func() {
+			wg.Wait()
+			close(ch)
+		}()
+
+		msg, ok := <-ch
+		assert.True(t, ok)
+		assert.NotNil(t, msg)
+		assert.IsType(t, []ec2Types.Volume{}, msg)
+	})
+}
+
+func Test_EmitMetricsFromVolumesChannel(t *testing.T) {
+	t.Run("reads from volumes channel and sends it over to prometheus channel", func(t *testing.T) {
+		volumesCh := make(chan []ec2Types.Volume)
+		promCh := make(chan prometheus.Metric)
+
+		regionName := "af-south-1"
+		region := ec2Types.Region{
+			RegionName: aws.String(regionName),
+		}
+		volumeType := "gp3"
+
+		ps := mockpricing.NewPricing(t)
+		collector := New(&Config{
+			Regions: []ec2Types.Region{region},
+			Logger:  logger,
+		}, ps)
+
+		collector.storagePricingMap = &StoragePricingMap{
+			Regions: map[string]*StoragePricing{
+				regionName: {
+					Storage: map[string]float64{
+						volumeType: 0.1047,
+					},
+				},
+			},
+		}
+
+		originMsg := []ec2Types.Volume{
+			{
+				AvailabilityZone: aws.String(fmt.Sprintf("%sa", regionName)),
+				VolumeId:         aws.String("vol-111111111"),
+				VolumeType:       ec2Types.VolumeType(volumeType),
+				Size:             aws.Int32(100),
+			},
+		}
+
+		// fill volumes channel with data from the above volume
+		wg := sync.WaitGroup{}
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			volumesCh <- originMsg
+			wg.Wait()
+			close(volumesCh)
+		}()
+
+		wg2 := sync.WaitGroup{}
+		wg2.Add(1)
+		go func() {
+			defer wg2.Done()
+			collector.emitMetricsFromVolumesChannel(volumesCh, promCh)
+			wg2.Wait()
+		}()
```

Comment:
WaitGroups were the only way I found to make these tests work. I know very little about channels and wait groups, but I was getting stuck on here, waiting for something on the channel, and the tests timed out. Happy to refactor/review with someone who can provide some insight into it.

Reply:
this one I'm also a bit stumped on...